File: swift/Sources/Wallet/PresentationService.swift
import CborCose
import CredentialStore
import Foundation
import MDoc
import OpenID4VP
import SdJwt
import StatusList
import TransactionLog
import Trust
import WalletAPI

/// OpenID4VP remote presentation. Bridges the store to the VP engine + records audit.
public struct PresentationService {
    let vp: Openid4VpClient
    let store: DefaultCredentialStore
    let txlog: TransactionLog
    let secureAreas: [any SecureArea]
    /// Registrar-scoped Token Status List client for RP registration certs (WRPRC); nil when no registrar
    /// anchors are configured. Used to refuse a revoked WRPRC before the consent screen.
    var registrarStatusClient: StatusListClient?
    /// Registrar TS5 API client for the dataset-only path (no WRPRC); nil when no registrar anchors are
    /// configured. Consulted only when `verifyRegistrationViaApi` is on (RPRC_16).
    var registrarApi: RegistrarApiClient?
    /// RPRC_16 opt-in: consult `registrarApi` to confirm a dataset-only RP's registration before consent.
    var verifyRegistrationViaApi: Bool = false
    /// When true, a failed final submission is recorded with `.error` status (opt-in via config).
    var recordFailures: Bool = false
    /// ISO 18013-5 §9.1.3.5 device-auth preference for mdoc presentations (deviceMac when the verifier requests it).
    var deviceAuthMode: MdocDeviceAuthMode = .signature
    var transactionDataBinder: OpenID4VP.MdocTransactionDataBinder?

    /// Remote (URL/QR) presentation: resolve → match stored credentials → consent → direct_post submit.
    public func start(_ requestUri: String) -> PresentationSession {
        runSession(
            resolve: { try await catchingVp { try await vp.resolveRequest(requestUri) } },
            submit: { resolved, matches, selection, held in
                let result = try await catchingVp {
                    try await vp.respond(request: resolved, matches: matches, selection: toVpSelection(selection), held: held)
                }
                return .completed(redirectUri: result.redirectUri, dcApiResponse: nil)
            })
    }

    /// Digital Credentials API presentation (browser-mediated). The platform hands over the `requestObject`
    /// and the caller `origin`; no HTTP is performed — the response object is returned in
    /// `PresentationState.completed(dcApiResponse:)` for the app to pass back to the platform.
    public func startDcApi(_ requestObject: String, origin: String) -> PresentationSession {
        runSession(
            resolve: { try await catchingVp { try await vp.resolveDcApiRequest(requestObject, origin: origin) } },
            submit: { resolved, matches, selection, held in
                let response = try await catchingVp {
                    try await vp.respondDcApi(request: resolved, matches: matches, selection: toVpSelection(selection), held: held)
                }
                return .completed(redirectUri: nil, dcApiResponse: response.serialize())
            })
    }

    private func runSession(
        resolve: @escaping () async throws -> ResolvedRequest,
        submit: @escaping (ResolvedRequest, DcqlMatchResult, PresentationSelection, [any PresentableCredential]) async throws -> PresentationState
    ) -> PresentationSession {
        let session = PresentationSession { s in
            s.emit(.resolvingRequest)
            let resolved = try await resolve()
            // Refuse a revoked RP registration cert (WRPRC) before doing any matching / consent. Only runs
            // when the request actually carried a WRPRC (registration != nil); absent → nothing to check.
            let statusValid = await checkRegistrationStatus(resolved)
            // Dataset-only path (no WRPRC): if the User opted in, confirm the RP's registration against the
            // registrar's TS5 API (RPRC_18). Best-effort — a failure leaves the self-declared dataset unverified.
            let registrarVerifiedCreds = await resolveRegistrarApi(resolved)

            let envelopes = try await store.list().filter { if case .issued = $0.lifecycle { return true }; return false }
            var held: [any PresentableCredential] = []
            for envelope in envelopes {
                if let p = presentableFor(envelope, firstInstance(envelope)) { held.append(p) }
            }
            let matches = vp.match(resolved, held: held)
            let request = buildRequest(resolved, matches, registrarVerifiedCreds: registrarVerifiedCreds, statusValid: statusValid)

            switch await s.awaitDecision(request) {
            case .none:
                try await recordDeclined(resolved)
                // §8.5: tell the verifier the user refused so it stops waiting. Best-effort — an
                // unreachable verifier must not turn a decline into a failure. DC API has no
                // response_uri; there the platform surfaces the refusal (§15.9.2).
                var redirectUri: String?
                if resolved.responseUri != nil {
                    redirectUri = try? await vp.reportError(
                        resolved, code: .accessDenied,
                        description: "the user declined to share the requested credentials"
                    ).redirectUri
                }
                s.emit(.declined(redirectUri: redirectUri))
            case let .some(selection):
                if selection.chosen.isEmpty { throw PresentationError.selectionIncomplete("no credential selected") }
                s.emit(.submitting)
                let chosenHeld = try await buildChosenHeld(envelopes, selection)
                let terminal: PresentationState
                do {
                    terminal = try await submit(resolved, matches, selection, chosenHeld)
                } catch {
                    // Only the final submission failed — record the attempted disclosure with .error status (opt-in).
                    if recordFailures { try? await recordError(resolved, selection, matches, statusValid) }
                    throw error
                }
                try await recordSuccess(resolved, selection, matches, statusValid)
                s.emit(terminal)
            }
        }
        session.launch()
        return session
    }

    /// Dataset-only registration confirmation (wrprc.md §5, RPRC_16/18). Returns the RP's registrar-signed
    /// registered credentials when the request carried only a self-declared `registrar_dataset` AND the User
    /// opted in AND the TS5 lookup succeeds; nil otherwise (WRPRC-attested, opted out, or the call failed — in
    /// which case we proceed with the self-declared dataset, unverified, per §5.3).
    private func resolveRegistrarApi(_ resolved: ResolvedRequest) async -> [RegisteredCredential]? {
        guard let reg = resolved.verifier.registration else { return nil }
        if reg.attested { return nil }                 // a WRPRC already gives authoritative, offline-verified registration
        guard verifyRegistrationViaApi, let api = registrarApi else { return nil } // RPRC_16 opt-in
        guard let registryURI = reg.dataset?.registryURI, let identifier = reg.dataset?.identifier else { return nil }
        // §5.3: could not obtain the registered info → proceed, the dataset stays unverified.
        return try? await api.fetchRegisteredCredentials(registryURI: registryURI, identifier: identifier,
                                                         intendedUseIdentifier: reg.dataset?.intendedUseIdentifier)
    }

    private func buildRequest(_ resolved: ResolvedRequest, _ matches: DcqlMatchResult,
                              registrarVerifiedCreds: [RegisteredCredential]? = nil,
                              statusValid: Bool? = nil) -> PresentationRequest {
        let required = matches.requiredQueryIds
        let queries = matches.candidatesByQuery.map { queryId, candidates in
            QueryPresentation(
                queryId: queryId, required: required.contains(queryId),
                candidates: candidates.map { PresentationCandidate(credentialId: CredentialId($0.credential.credentialId), disclosedPaths: $0.disclosedPaths) },
                multiple: candidates.first?.query.multiple ?? false)
        }
        let v = resolved.verifier
        let registration = v.registration.map { r -> VerifierRegistration in
            // Prefer the registrar-signed credentials from the TS5 lookup (dataset-only + opt-in) over the
            // self-declared ones; a WRPRC-attested request already carries authoritative registeredCredentials.
            let registered = registrarVerifiedCreds ?? r.registeredCredentials
            // RPRC_21 attribute-scope check: which requested attributes fall outside what the RP registered.
            let unregistered = RegistrationScope.unregistered(resolved.dcqlQuery, registered: registered)
            return VerifierRegistration(
                subject: r.subject, entitlements: r.entitlements,
                purpose: r.purpose.map { PurposeText(lang: $0.lang, value: $0.value) },
                intermediarySub: r.intermediarySub, intermediaryName: r.intermediaryName,
                // Reaching here means any revoked WRPRC was already refused; validated iff a status client ran.
                statusValid: statusValid,
                attested: r.attested,
                registrarVerified: r.attested || registrarVerifiedCreds != nil,
                registryURI: r.dataset?.registryURI,
                policyURI: r.dataset?.policyURI,
                unregisteredClaims: unregistered.map { $0.path })
        }
        return PresentationRequest(
            verifier: VerifierInfo(clientId: v.clientId, clientIdScheme: v.clientIdScheme, commonName: v.commonName,
                                   trusted: v.trusted, registration: registration),
            queries: queries, transactionData: resolved.transactionData, satisfiable: matches.isSatisfiable(),
            resolved: resolved, matches: matches)
    }

    /// Refuses a revoked/suspended RP registration cert (WRPRC) via the Token Status List, when the request
    /// carried a WRPRC (`resolved.verifier.registration`) and a registrar status client is configured. A
    /// missing WRPRC or missing status client is a no-op — interop with verifiers that don't send one yet.
    /// The RP registration cert (WRPRC) Token Status List result, surfaced (not enforced): true = valid,
    /// false = revoked/suspended (or the check failed), nil = nothing to check. The wallet shows it and the
    /// User decides — a revoked verifier does not hard-fail the presentation.
    private func checkRegistrationStatus(_ resolved: ResolvedRequest) async -> Bool? {
        guard let status = resolved.verifier.registration?.status, let client = registrarStatusClient else { return nil }
        do {
            return try await client.check(claims: .obj([("status", status)])) == .valid
        } catch {
            return false // could not confirm the status → treat as not-valid, surfaced to the User (not a hard fail)
        }
    }

    /// Consumes one instance per chosen credential (usage counting) and builds a signer-backed presentable.
    private func buildChosenHeld(_ envelopes: [CredentialEnvelope], _ selection: PresentationSelection) async throws -> [any PresentableCredential] {
        var byId: [String: CredentialEnvelope] = [:]
        for envelope in envelopes { byId[envelope.id.value] = envelope }
        var result: [any PresentableCredential] = []
        for idValue in Set(selection.chosen.values.flatMap { $0 }.map { $0.value }) {
            guard let envelope = byId[idValue] else { continue }
            guard let consumed = try await store.consumeInstance(CredentialId(idValue)) else { continue }
            if let p = presentableFor(envelope, consumed.instance) { result.append(p) }
        }
        return result
    }

    private func presentableFor(_ envelope: CredentialEnvelope, _ instance: CredentialInstance?) -> (any PresentableCredential)? {
        guard let instance else { return nil }
        guard let area = secureAreas.first(where: { $0.id == instance.key.secureArea }) else { return nil }
        do {
            switch envelope.format {
            case .sdJwtVc:
                return try HeldSdJwtVc(
                    credentialId: envelope.id.value,
                    sdJwt: try SdJwt.parse(String(decoding: instance.payload, as: UTF8.self)),
                    holderSigner: SecureAreaJwsSigner(area: area, key: instance.key, algorithm: .es256))
            case .msoMdoc:
                let key = instance.key
                var keyAgreement: MdocKeyAgreement?
                if area.capabilities.keyAgreement {
                    keyAgreement = { peer in try await area.keyAgreement(key: key, peerPublicKey: peer, hint: nil) }
                }
                return try HeldMdoc(
                    credentialId: envelope.id.value,
                    issuerSigned: try IssuerSigned.decode(instance.payload),
                    deviceSigner: SecureAreaCoseSigner(area: area, key: key, algorithm: .es256),
                    deviceKeyAgreement: keyAgreement,
                    deviceAuth: deviceAuthMode,
                    transactionDataBinder: transactionDataBinder)
            }
        } catch {
            return nil
        }
    }

    private func toVpSelection(_ selection: PresentationSelection) -> OpenID4VP.PresentationSelection {
        var chosen: [String: [String]] = [:]
        for (queryId, credentialIds) in selection.chosen { chosen[queryId] = credentialIds.map { $0.value } }
        return OpenID4VP.PresentationSelection(chosen: chosen)
    }

    private func firstInstance(_ envelope: CredentialEnvelope) -> CredentialInstance? {
        if case let .issued(_, instances) = envelope.lifecycle { return instances.first }
        return nil
    }

    private func recordSuccess(_ resolved: ResolvedRequest, _ selection: PresentationSelection, _ matches: DcqlMatchResult, _ statusValid: Bool?) async throws {
        await txlog.recordPresentation(relyingParty: relyingPartyOf(resolved, statusValid), documents: loggedDocuments(selection, matches), status: .success, transport: .remote)
    }

    private func recordDeclined(_ resolved: ResolvedRequest, _ statusValid: Bool? = nil) async throws {
        await txlog.recordPresentation(relyingParty: relyingPartyOf(resolved, statusValid), documents: [], status: .incomplete, transport: .remote)
    }

    private func recordError(_ resolved: ResolvedRequest, _ selection: PresentationSelection, _ matches: DcqlMatchResult, _ statusValid: Bool?) async throws {
        await txlog.recordPresentation(relyingParty: relyingPartyOf(resolved, statusValid), documents: loggedDocuments(selection, matches), status: .error, transport: .remote)
    }

    private func relyingPartyOf(_ resolved: ResolvedRequest, _ statusValid: Bool?) -> RelyingParty {
        let v = resolved.verifier
        let reg = v.registration
        return RelyingParty(
            id: v.clientId,
            name: v.commonName,
            trusted: v.trusted,
            certificateChainDer: v.certificateChainDer ?? [],
            clientIdScheme: v.clientIdScheme,
            subject: reg?.subject,
            entitlements: reg?.entitlements ?? [],
            purpose: reg?.purpose.map { LocalizedText(lang: $0.lang, value: $0.value) } ?? [],
            intermediaryName: reg?.intermediaryName,
            intermediarySub: reg?.intermediarySub,
            attested: reg?.attested,
            statusValid: statusValid)
    }

    private func loggedDocuments(_ selection: PresentationSelection, _ matches: DcqlMatchResult) -> [LoggedDocument] {
        selection.chosen.flatMap { queryId, credentialIds in
            credentialIds.compactMap { credentialId -> LoggedDocument? in
                guard let candidate = matches.candidatesByQuery[queryId]?.first(where: { $0.credential.credentialId == credentialId.value }) else { return nil }
                return LoggedDocument(
                    format: candidate.credential.format,
                    type: candidate.credential.vct ?? candidate.credential.docType,
                    queryId: queryId,
                    claims: candidate.disclosedPaths.map { LoggedClaim(path: $0) })
            }
        }
    }

    private func catchingVp<T>(_ block: () async throws -> T) async throws -> T {
        do {
            return try await block()
        } catch let e as TrustError {
            throw PresentationError.verifierNotTrusted(e.description)
        } catch VpError.invalidRequest(let m) {
            throw PresentationError.invalidRequest(m)
        } catch VpError.verifierNotTrusted(let m) {
            throw PresentationError.verifierNotTrusted(m)
        } catch VpError.queryNotSatisfiable(let missing) {
            throw PresentationError.queryNotSatisfiable("\(missing)")
        } catch VpError.selectionIncomplete(let m) {
            throw PresentationError.selectionIncomplete(m)
        } catch VpError.responseFailed(let m) {
            throw PresentationError.responseRejected(m)
        } catch VpError.unsupported(let m) {
            throw PresentationError.invalidRequest(m)
        }
    }
}

File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/PresentationService.kt
package com.hopae.eudi.wallet

import com.hopae.eudi.wallet.mdoc.IssuerSigned
import com.hopae.eudi.wallet.mdoc.MdocDeviceAuthMode
import com.hopae.eudi.wallet.mdoc.MdocKeyAgreement
import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.sdjwt.SdJwt
import com.hopae.eudi.wallet.sdjwt.SecureAreaJwsSigner
import com.hopae.eudi.wallet.spi.CredentialFormat
import com.hopae.eudi.wallet.spi.CredentialId
import com.hopae.eudi.wallet.spi.SecureArea
import com.hopae.eudi.wallet.spi.SecureAreaCoseSigner
import com.hopae.eudi.wallet.spi.SigningAlgorithm
import com.hopae.eudi.wallet.status.CredentialStatus as StatusListStatus
import com.hopae.eudi.wallet.status.StatusListClient
import com.hopae.eudi.wallet.status.StatusListException
import com.hopae.eudi.wallet.store.CredentialEnvelope
import com.hopae.eudi.wallet.store.CredentialInstance
import com.hopae.eudi.wallet.store.CredentialStore
import com.hopae.eudi.wallet.store.EnvelopeLifecycle
import com.hopae.eudi.wallet.trust.TrustException
import com.hopae.eudi.wallet.txlog.LoggedClaim
import com.hopae.eudi.wallet.txlog.LoggedDocument
import com.hopae.eudi.wallet.txlog.RelyingParty
import com.hopae.eudi.wallet.txlog.TransactionLog
import com.hopae.eudi.wallet.txlog.TransactionStatus
import com.hopae.eudi.wallet.vp.DcqlMatchResult
import com.hopae.eudi.wallet.vp.HeldMdoc
import com.hopae.eudi.wallet.vp.HeldSdJwtVc
import com.hopae.eudi.wallet.vp.Openid4VpClient
import com.hopae.eudi.wallet.vp.PresentableCredential
import com.hopae.eudi.wallet.vp.RegistrationScope
import com.hopae.eudi.wallet.vp.ResolvedRequest
import com.hopae.eudi.wallet.vp.VpErrorCode
import com.hopae.eudi.wallet.vp.VpException
import kotlinx.coroutines.CoroutineScope
import com.hopae.eudi.wallet.vp.PresentationSelection as VpSelection

/** OpenID4VP remote presentation. Bridges the store to the VP engine + records audit. */
class PresentationService internal constructor(
    private val vp: Openid4VpClient,
    private val store: CredentialStore,
    private val txlog: TransactionLog,
    private val secureAreas: List<SecureArea>,
    private val scope: CoroutineScope,
    /**
     * Registrar-scoped Token Status List client for RP registration certs (WRPRC); null when no registrar
     * anchors are configured. Used to refuse a revoked WRPRC before the consent screen.
     */
    private val registrarStatusClient: StatusListClient? = null,
    /** When true, a failed final submission is recorded with ERROR status (opt-in via config). */
    private val recordFailures: Boolean = false,
    /** ISO 18013-5 §9.1.3.5 device-auth preference for mdoc presentations (deviceMac when the verifier requests it). */
    private val deviceAuthMode: MdocDeviceAuthMode = MdocDeviceAuthMode.Signature,
    /** Host mapping for mdoc `transaction_data` device-signed elements (ISO 18013-7 B.2.1); null = reject. */
    private val transactionDataBinder: com.hopae.eudi.wallet.vp.MdocTransactionDataBinder? = null,
) {
    /** Remote (URL/QR) presentation: resolve → match stored credentials → consent → direct_post submit. */
    fun start(requestUri: String): PresentationSession = runSession(
        resolve = { catchingVp { vp.resolveRequest(requestUri) } },
        submit = { resolved, matches, selection, held ->
            PresentationState.Completed(catchingVp { vp.respond(resolved, matches, toVpSelection(selection), held) }.redirectUri)
        },
    )

    /**
     * Digital Credentials API presentation (browser-mediated). The platform hands over the [requestObject]
     * and the caller [origin]; no HTTP is performed — the response object is returned in
     * [PresentationState.Completed.dcApiResponse] for the app to pass back to the platform.
     */
    fun startDcApi(requestObject: String, origin: String): PresentationSession = runSession(
        resolve = { catchingVp { vp.resolveDcApiRequest(requestObject, origin) } },
        submit = { resolved, matches, selection, held ->
            PresentationState.Completed(
                redirectUri = null,
                dcApiResponse = catchingVp { vp.respondDcApi(resolved, matches, toVpSelection(selection), held) }.serialize(),
            )
        },
    )

    private fun runSession(
        resolve: suspend () -> ResolvedRequest,
        submit: suspend (ResolvedRequest, DcqlMatchResult, PresentationSelection, List<PresentableCredential>) -> PresentationState,
    ): PresentationSession {
        val session = PresentationSession(scope) {
            emit(PresentationState.ResolvingRequest)
            val resolved = resolve()
            // Refuse a revoked RP registration cert (WRPRC) before any matching / consent. Only runs when
            // the request actually carried a WRPRC (registration != null); absent → nothing to check.
            catchingVp { checkRegistrationStatus(resolved) }

            val envelopes = store.list().filter { it.lifecycle is EnvelopeLifecycle.Issued }
            val held = envelopes.mapNotNull { presentableFor(it, it.firstInstance()) }
            val matches = vp.match(resolved, held)
            val request = buildRequest(resolved, matches)

            when (val selection = awaitDecision(request)) {
                null -> {
                    recordDeclined(resolved)
                    // §8.5: tell the verifier the user refused so it stops waiting. Best-effort — an
                    // unreachable verifier must not turn a decline into a failure. DC API has no
                    // response_uri; there the platform surfaces the refusal (§15.9.2).
                    val redirectUri = resolved.responseUri?.let {
                        runCatching {
                            vp.reportError(resolved, VpErrorCode.ACCESS_DENIED, "the user declined to share the requested credentials")
                        }.getOrNull()?.redirectUri
                    }
                    emit(PresentationState.Declined(redirectUri))
                }
                else -> {
                    if (selection.chosen.isEmpty()) throw WalletError.Presentation.SelectionIncomplete("no credential selected")
                    emit(PresentationState.Submitting)
                    val chosenHeld = buildChosenHeld(envelopes, selection)
                    val terminal = try {
                        submit(resolved, matches, selection, chosenHeld)
                    } catch (e: Throwable) {
                        // Only the final submission failed — record the attempted disclosure with ERROR status (opt-in).
                        if (recordFailures) runCatching { recordError(resolved, selection, matches) }
                        throw e
                    }
                    recordSuccess(resolved, selection, matches)
                    emit(terminal)
                }
            }
        }
        session.launch()
        return session
    }

    private fun buildRequest(resolved: ResolvedRequest, matches: DcqlMatchResult): PresentationRequest {
        val required = matches.requiredQueryIds
        val queries = matches.candidatesByQuery.map { (queryId, candidates) ->
            QueryPresentation(
                queryId = queryId,
                required = queryId in required,
                candidates = candidates.map { PresentationCandidate(CredentialId(it.credential.credentialId), it.disclosedPaths) },
                multiple = candidates.firstOrNull()?.query?.multiple ?: false,
            )
        }
        val v = resolved.verifier
        val registration = v.registration?.let { r ->
            // RPRC_21 attribute-scope check: which requested attributes fall outside what the RP registered.
            val unregistered = RegistrationScope.unregistered(resolved.dcqlQuery, r.registeredCredentials)
            VerifierRegistration(
                subject = r.subject,
                entitlements = r.entitlements,
                purpose = r.purpose.map { PurposeText(it.lang, it.value) },
                intermediarySub = r.intermediarySub,
                intermediaryName = r.intermediaryName,
                // Reaching here means any revoked WRPRC was already refused; validated iff a status client ran.
                statusValid = if (r.status != null && registrarStatusClient != null) true else null,
                attested = r.attested,
                registryURI = r.dataset?.registryURI,
                policyURI = r.dataset?.policyURI,
                unregisteredClaims = unregistered.map { it.path },
            )
        }
        return PresentationRequest(
            verifier = VerifierInfo(v.clientId, v.clientIdScheme, v.commonName, v.trusted, registration),
            queries = queries,
            transactionData = resolved.transactionData,
            satisfiable = matches.isSatisfiable(),
            resolved = resolved,
            matches = matches,
        )
    }

    /**
     * Refuses a revoked/suspended RP registration cert (WRPRC) via the Token Status List, when the request
     * carried a WRPRC ([ResolvedRequest.verifier].registration) and a registrar status client is configured.
     * A missing WRPRC or missing status client is a no-op — interop with verifiers that don't send one yet.
     */
    private suspend fun checkRegistrationStatus(resolved: ResolvedRequest) {
        val status = resolved.verifier.registration?.status ?: return
        val client = registrarStatusClient ?: return
        val result = try {
            client.check(JsonValue.Obj(listOf("status" to status)))
        } catch (e: StatusListException) {
            throw VpException.VerifierNotTrusted("WRPRC status list check failed: ${e.message}")
        }
        if (result != StatusListStatus.VALID) {
            throw VpException.VerifierNotTrusted("WRPRC status is $result (revoked or suspended)")
        }
    }

    /** Consumes one instance per chosen credential (usage counting) and builds a signer-backed presentable. */
    private suspend fun buildChosenHeld(envelopes: List<CredentialEnvelope>, selection: PresentationSelection): List<PresentableCredential> {
        val byId = envelopes.associateBy { it.id }
        return selection.chosen.values.flatten().distinct().mapNotNull { credentialId ->
            val envelope = byId[credentialId] ?: return@mapNotNull null
            val consumed = store.consumeInstance(credentialId) ?: return@mapNotNull null
            presentableFor(envelope, consumed.instance)
        }
    }

    private fun presentableFor(envelope: CredentialEnvelope, instance: CredentialInstance?): PresentableCredential? {
        if (instance == null) return null
        val area = secureAreas.firstOrNull { it.id == instance.key.secureArea } ?: return null
        return runCatching {
            when (envelope.format) {
                is CredentialFormat.SdJwtVc -> HeldSdJwtVc(
                    envelope.id.value,
                    SdJwt.parse(instance.payload.decodeToString()),
                    SecureAreaJwsSigner(area, instance.key, SigningAlgorithm.ES256),
                )
                is CredentialFormat.MsoMdoc -> HeldMdoc(
                    envelope.id.value,
                    IssuerSigned.decode(instance.payload),
                    SecureAreaCoseSigner(area, instance.key, SigningAlgorithm.ES256),
                    deviceKeyAgreement = if (area.capabilities.keyAgreement) {
                        MdocKeyAgreement { peer -> area.keyAgreement(instance.key, peer) }
                    } else null,
                    deviceAuth = deviceAuthMode,
                    transactionDataBinder = transactionDataBinder,
                )
            }
        }.getOrNull()
    }

    private fun toVpSelection(selection: PresentationSelection): VpSelection =
        VpSelection(selection.chosen.mapValues { (_, ids) -> ids.map { it.value } })

    private fun CredentialEnvelope.firstInstance(): CredentialInstance? =
        (lifecycle as? EnvelopeLifecycle.Issued)?.instances?.firstOrNull()

    private suspend fun recordSuccess(resolved: ResolvedRequest, selection: PresentationSelection, matches: DcqlMatchResult) {
        txlog.recordPresentation(relyingPartyOf(resolved), loggedDocuments(selection, matches), TransactionStatus.SUCCESS)
    }

    private suspend fun recordDeclined(resolved: ResolvedRequest) {
        txlog.recordPresentation(relyingPartyOf(resolved), documents = emptyList(), status = TransactionStatus.INCOMPLETE)
    }

    private suspend fun recordError(resolved: ResolvedRequest, selection: PresentationSelection, matches: DcqlMatchResult) {
        txlog.recordPresentation(relyingPartyOf(resolved), loggedDocuments(selection, matches), TransactionStatus.ERROR)
    }

    private fun relyingPartyOf(resolved: ResolvedRequest): RelyingParty = RelyingParty(
        id = resolved.verifier.clientId,
        name = resolved.verifier.commonName,
        trusted = resolved.verifier.trusted,
        certificateChainDer = resolved.verifier.certificateChainDer ?: emptyList(),
    )

    private fun loggedDocuments(selection: PresentationSelection, matches: DcqlMatchResult): List<LoggedDocument> =
        selection.chosen.flatMap { (queryId, credentialIds) ->
            credentialIds.mapNotNull { credentialId ->
                val candidate = matches.candidatesByQuery[queryId].orEmpty()
                    .firstOrNull { it.credential.credentialId == credentialId.value } ?: return@mapNotNull null
                LoggedDocument(
                    format = candidate.credential.format,
                    type = candidate.credential.vct ?: candidate.credential.docType,
                    queryId = queryId,
                    claims = candidate.disclosedPaths.map { LoggedClaim(it, null) },
                )
            }
        }

    private suspend fun <T> catchingVp(block: suspend () -> T): T = try {
        block()
    } catch (e: TrustException) {
        throw WalletError.Presentation.VerifierNotTrusted(e.message ?: "reader certificate not trusted")
    } catch (e: VpException) {
        throw when (e) {
            is VpException.InvalidRequest -> WalletError.Presentation.InvalidRequest(e.message ?: "", e)
            is VpException.VerifierNotTrusted -> WalletError.Presentation.VerifierNotTrusted(e.message ?: "")
            is VpException.SelectionIncomplete -> WalletError.Presentation.SelectionIncomplete(e.message ?: "")
            is VpException.ResponseFailed -> WalletError.Presentation.ResponseRejected(e.message ?: "", e)
            is VpException.Unsupported -> WalletError.Presentation.InvalidRequest(e.message ?: "", e)
            else -> WalletError.Presentation.Unexpected(e)
        }
    }
}
